File: apps/modernization-ui/src/apps/patient/file/events/reports/laboratory/patientLabReport.ts
import { AssociatedInvestigation } from 'libs/events/investigations/associated';
import { Specimen } from 'libs/events/reports/laboratory';
import { ResultedTest } from 'libs/events/tests';
import { Provider } from 'libs/provider';

type PatientLabReport = {
    patient: number;
    id: number;
    local: string;
    programArea: string;
    jurisdiction: string;
    receivedDate?: Date;
    electronic: boolean;
    processingDecision?: string;
    collectedDate?: Date;
    resultedTests?: ResultedTest[];
    reportingFacility?: string;
    orderingProvider?: Provider;
    orderingFacility?: string;
    specimen?: Specimen;
    associations?: AssociatedInvestigation[];
};

export type { PatientLabReport };

File: apps/modernization-api/src/main/java/gov/cdc/nbs/patient/file/events/report/laboratory/PatientLabReport.java
package gov.cdc.nbs.patient.file.events.report.laboratory;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import gov.cdc.nbs.demographics.name.DisplayableSimpleName;
import gov.cdc.nbs.patient.events.investigation.association.AssociatedInvestigation;
import gov.cdc.nbs.patient.events.tests.ResultedTest;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Collection;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record PatientLabReport(
    @JsonProperty(required = true) long patient,
    @JsonProperty(required = true) long id,
    @JsonProperty(required = true) String local,
    @JsonProperty(required = true) String programArea,
    @JsonProperty(required = true) String jurisdiction,
    LocalDateTime receivedDate,
    @JsonProperty(required = true)
    boolean electronic,
    String processingDecision,
    LocalDate collectedDate,
    Collection<ResultedTest> resultedTests,
    String reportingFacility,
    DisplayableSimpleName orderingProvider,
    String orderingFacility,
    Specimen specimen,
    Collection<AssociatedInvestigation> associations
) {

  public record Specimen(String site, String source) {
  }



  PatientLabReport withResultedTests(final Collection<ResultedTest> resultedTests) {
    return new PatientLabReport(
        patient(),
        id(),
        local(),
        programArea(),
        jurisdiction(),
        receivedDate(),
        electronic(),
        processingDecision(),
        collectedDate(),
        resultedTests,
        reportingFacility(),
        orderingProvider(),
        orderingFacility(),
        specimen(),
        associations()
    );
  }

  PatientLabReport withAssociations(final Collection<AssociatedInvestigation> associations) {
    return new PatientLabReport(
        patient(),
        id(),
        local(),
        programArea(),
        jurisdiction(),
        receivedDate(),
        electronic(),
        processingDecision(),
        collectedDate(),
        resultedTests(),
        reportingFacility(),
        orderingProvider(),
        orderingFacility(),
        specimen(),
        associations
    );
  }
}
